feat:  Adds in ParseOption classes

Zemill opened the pull request:
This class along with its associated code allows the users to set different options for parsing.  Some default options are already defined (minimal, typical, and full)

The purpose of this change is to reduce cycle time of the parser parsing data that is not used by the end user.  Things like mouse events and messages are removed from the Typical ParseOptions option.

Some minor syntax changes.  These  changes were originally written by  @poma and @martijnhoekstra and consolidated by @zemill

Zemill committed:
feat:  Adds in ParseOption classes

This class along with its associated code allows the users to set different options for parsing.  Some default options are already defined (minimal, typical, and full)

The purpose of this change is to reduce cycle time of the parser parsing data that is not used by the end user.  Things like mouse events and messages are removed from the Typical ParseOptions option.

Some minor syntax changes.  These  changes were originally written by  @poma and @martijnhoekstra and consolidated by @zemill

Zemill committed:
fix: added tracker events back in

barrett777 commented:
Thanks for putting this together!  Could you resolve the merge conflicts?

Zemill committed:
Fix:  Resolved conflict on program.cs

Zemill committed:
fix: resolved conflict on BitReader.cs file

Zemill committed:
fix: Resolved conflict on ReplayGameEvents.cs file

Zemill committed:
feat: Switched to Full Parsing as default

Zemill committed:
Merge branch 'master' into master

Zemill commented:
Ok. should be good to go now

Zemill committed:
fix: cleanup

barrett777 commented on Heroes.ReplayParser/ParseOptions.cs:
```diff
@@ -0,0 +1,56 @@
+        public bool ShouldParseMouseEvents { get; set; } = false;
+        public bool ShouldParseDetailedBattleLobby { get; set; } = true;
+        public bool ShouldParseMessageEvents { get; set; } = false;
+        public bool ShouldParseStatistics { get; set; } = true;
```
Looks good overall :) Yes I've needed an options class like this for awhile

One request: Could we set the defaults to match what the current defaults are, to hopefully avoid some confusion.  Users can of course specify whatever they want

The change would be to enable 'ShouldParseUnits', enable 'ShouldParseMouseEvents', and disable 'ShouldParseDetailedBattleLobby'

Zemill replied:
I have committed some changes to work with that.  I have removed TypicalParsing and replaced it with DefaultParsing.  I then created MediumParsing to capture what was previous TypicalParsing.  I also set the default parsing mode to DefaultParsing.

Take a look and let me know if that looks right

Zemill committed:
fix:  Updated ParseOptions to use original Default options

Updated ParseOptions to use original Default options and also set the default parse to DefaultParsing
